crates/typst-macros/src/elem.rs

```diff
@@ -480,88 +472,904 @@ fn create_param_info(field: &Field) -> TokenStream {
+    let needs_preparation = element
+        .without_capability("Unlabellable", || {
+            element
+                .capabilities
+                .iter()
+                .any(|capability| capability == "Locatable" || capability == "Synthesize")
+                .then(|| quote! { !self.prepared })
+                .unwrap_or_else(|| quote! { self.label().is_some() && !self.prepared })
+        })
+        .unwrap_or_else(|| {
+            assert!(element.capabilities.iter().all(|capability| capability
+                != "Locatable"
+                && capability != "Synthesize"));
+            quote! { false }
+        });
+
+    // Creation of the fields dictionary for inherent fields.
+    let field_dict = element.inherent_fields().clone().map(|field| {
+        let name = &field.name;
+        let field_ident = &field.ident;
+
+        quote! {
+            fields.insert(
+                EcoString::inline(#name).into(),
+                ::typst::eval::IntoValue::into_value(self.#field_ident.clone())
+            );
         }
-    }
-}
+    });
 
-/// Create the element's `Set` implementation.
-fn create_set_impl(element: &Elem) -> TokenStream {
-    let ident = &element.ident;
-    let handlers = element
-        .fields
-        .iter()
-        .filter(|field| {
-            !field.external
-                && !field.synthesized
-                && field.settable()
-                && (!field.internal || field.parse.is_some())
-        })
+    // Creation of the fields dictionary for optional fields.
+    let field_opt_dict = element
+        .visible_fields()
+        .filter(|field| !field.inherent())
+        .clone()
         .map(|field| {
-            let set_ident = &field.set_ident;
-            let (prefix, value) = create_field_parser(field);
+            let name = &field.name;
+            let field_ident = &field.ident;
+
             quote! {
-                #prefix
-                if let Some(value) = #value {
-                    styles.set(Self::#set_ident(value));
+                if let Some(value) = &self.#field_ident {
+                    fields.insert(
+                        EcoString::inline(#name).into(),
+                        ::typst::eval::IntoValue::into_value(value.clone())
+                    );
                 }
             }
         });
 
+    let location = element
+        .without_capability("Unlabellable", || {
+            quote! {
+                self.location
+            }
+        })
+        .unwrap_or_else(|| quote! { None });
+
+    let set_location = element
+        .without_capability("Unlabellable", || {
+            quote! {
+                self.location = Some(location);
+            }
+        })
+        .unwrap_or_else(|| quote! { drop(location) });
+
+    let label = element
+        .without_capability("Unlabellable", || {
+            quote! {
+                self.label
+            }
+        })
```

Comment:
```suggestion
        .without_capability("Unlabellable", || quote! { self.label })
```
same in similar places